### DISCUSSION TITLE
Why does toga lack events for a large number of widgets other than Button?

### DISCUSSION THREAD

**Discussion by wanglifeng95** [beeware/toga#2141] [General]
In my spare time, I watched local APIs for Android, Windows, and Linux. I found that many widgets of Toga, such as Box, ImageView, Label, etc., may not have been able to add click or long press events, perhaps because Linux PyGobject is using Gtk3.0. In Gtk3.0, button can use signals to trigger events normally, but many other widgets can only be added to EventBox() to trigger events. That is, these widgets can have events,but they are no longer themselves,instead become EventBox objects. Therefore when Toga uses local APIs to add events to these widgets, it will cause problems, and ultimately, toga give up the events of those widgets, right?
But one difficulty with Python is mainly in mobile applications. If Toga's goal is to make Python apps run better on various platforms(desktops and mobile), The Desktop OS can be based on Python's built-in TKinder (which is converted to Toga code). TKinder is lightweight and stable, and it is well adapted to various system platforms on the desktop end. Then Toga can focus on the most promising intelligent mobile terminals in the future, and Toga can have a better set of code to unify the operation of various desktops and mobile platforms, and it is easier to maintain. Alternatively, Toga may follow the previous approach, and Toga's local API for Linux may be based on PYGobject GTK4.0 in the future, which is more user-friendly and allows widgets to easily add events without the need to use EventBox() to change themselves.
The above is my thoughts on Toga after studying it and the local APIs of various operaitting systems. I don't know if my ideas are appropriate. I really like Toga and hope it gets better and better.

**Comment by freakboy3742** [comment]
I'm not exactly sure I understand what the question is.

It feels like you've asking for a feature that is currently missing... but I can't work out from your description what that need actually *is*. It might be *something* to do with mouse press events... but you haven't really explained what it is that you want to do that you can't do.

If you're comparing Toga with Tkinter... I'd argue that Toga is already *superior* to Tkinter. Tkinter doesn't have a web widget, or a canvas widget; and it's not *ever* likely to be available on iOS, Android, web, or text console - features that Toga has *right now*. Are there things you can do in Tkinter that you can't do in Toga? Sure. But I can tell you from experience that it will be a *lot* easier to add missing features to Toga than to add features than it will be to add missing features to Tkinter.

If your suggestion is that Toga should be made a wrapper around GTK4 on *all* platforms... I guess you've never used GTK on any platform other that Linux. In my experience, GTK apps look *awful* on macOS, are only marginally better on Windows - and won't work on iOS, Android, web, or text console. The only one of those platforms that is even plausible as a future GTK target is web (and *maybe* Android). So this is a non-starter as well.

If you've got a feature request, can I ask that you actually describe the feature request, rather than theorising about potential fantasy rewrites of Toga?

**Reply by wanglifeng95** [reply]
Sorry, I didn't describe it clearly before. It's actually curiosity, not questioning. When I was looking at the Linux local API (PyObject GTK3), I found that its widgets and event implementations were similar to Toga in some aspects. Therefore, I was curious whether Toga's original design was also influenced by GTK3, and whether Toga had any plans to upgrade the Linux local API Gtk3.0 to GTK4.0.

Based on the local API, I can implement so many function on Toga. Therefore, it was my further developed to write event methods so that Toga widgets can directly add events through it, which is common on Windows and Android. When I wanted to extend this method to Linux, I realized a significant flaw in GTK3. Various widgets , such as Label, are still Label objects after adding events on Windows and Android. However, when events are added on Linux Gtk3.0, they are no longer Label objects, but instead EventBox objects, making the label_ Impl.native returns an event label that is restricted.Other widgets also have the same problem, so cannot write a universal event method, at least for Linux special cases. It is precisely because of this and other questions that I deeply understand the difficulty of Toga in implementing a set of code that is compatible with various platforms, and understand why it abandoned the implementation of some functions, such as events on some widgets. 
You are right about Tkinter.The GTK4.0 mentioned earlier refers to Toga's upgrade of Linux's local API from GTK3.0 to GTK4.0, not saying that other platforms such as Mac, iPhone, Android, Windows, etc. also using GTK4.0.

**Reply by freakboy3742** [reply]
I wouldn't say Toga's design has any particular design influence from GTK, beyond the things that are common to *all* GUI toolkits.

We *do* have plans to update to GTK4 - see #1978 for details on that work. However, this work isn't an especially high priority at present, mostly because GTK4 isn't as widely available as GTK3.

As for the "significant flaw" you've described - I'm afraid I don't understand what you're saying. I don't know what a "universal" event method means, or what "restriction" you're referring to. We've got an entire widget toolkit that demonstrates that you can define a public API that describes a set of events, and then define implementations of those events that are platform specific; and I definitely haven't found anything fundamentally different about GTK that suggests there's some class of event that *can't* be implemented.

So - lets try this again: What *specific* event handler would you like to see in the Toga API that does not currently exist?

**Reply by wanglifeng95** [reply]
I like to see click events, long press events, or right-click events on widgets such as ImageView, Box, Label, etc in the Toga.Many people are accustomed to separating UI and action functions when developing an app, regarding the issue of Linux Gtk3.0, I take this as an example：

```
#Implementing a user interface
class ChatBox(toga.Box):
    def __init__(self,*args,**kwargs):
        self.label=toga.Label("label")
        self.add(label)
        ...

#Implement its actions and functions
class ChatBoxAction:
    def __init__(self):
        self.chat=ChatBox()
        #Bind events (such as clicking, long-pressing, or right-clicking) and implement interactive functions for widgets such as Button, 
         #ImageView, Box, and Label in ChatBox here.
         self.clickEvent(self.chat.label)    #label add click event
         ....

    def clickEvent(self,widget)
       labelNative=widget._impl.native
       #It binds click event through local APIs, the widget implements a certain event
       ....
```

The above code originally looked comfortable and smooth in both Windows and Android,just call ChatBoxAction().chat.
But this cannot be done in Linux.
In Linux, when I need to add events to a widget such as label, I have to do so：
```
    labelNative=label._impl.native
    labelEventBox=EventBox()
    labelEventBox.add(labelNative)
    labelEventBox.connect('signal',hanler)
```
In linux, every time a widget want to add an event, I have to change the UI class to add a eventBox.The code becomes easily coupled, making it difficult to separate interface design and functional implementation. At the same time, the code also appears inconsistent with Windows and Android.Gtk4.0 may also not like this way of changing existing widgets, so it abandoned EventBox.So, I'm glad to see that Toga has an upgrade plan of Linux's local API from GTK3.0 to GTK4.0，although it isn't an especially high priority at presen.

**Reply by mhsmith** [reply]
There's already discussion of making a clickable ImageView at #774, and a clickable Label at #1022.

**Reply by freakboy3742** [reply]
We are unlikely to add support for click events to Label or Box. 

From a UX design perspective, there isn't a good reason for these UI elements to *be* clickable, as their appearance provides no indication that they're inherently clickable. If you want something to respond like a button... use a button. That's what a button *is*. The visual design of a button indicates to the user that it *can* be pressed.

This is fundamentally why the API for clickable labels on GTK seems odd to you. Labels *aren't* clickable. A GUI *shouldn't* be designed to require that kind of interaction. The "workaround" you've found is to wrap the label in a widget that *is* clickable - an EventBox. That's a *very* GTK-specific workaround, but it is indicative of a general philosophy of GUI design. macOS, iOS, and Windows don't provide a click handler for Labels either - but workarounds also exist on those platforms.

Adding a click handler to ImageView is a more complex choice. However, my inclination is to strongly resist the temptation to add this - because again, *an image isn't meant to be clickable*. An image provides no indication that it *is* clickable. The subject of "clickable images" has come up many, many times in the past - see #774, #1343... and #1959, the last time *you* asked this question.

In every instance that someone has requested an "clickable image", what they are *actually* asking for is:

1. A button with an *icon* on it. e.g., a "stop" button that is a red stop sign, rather than the text "Stop". Ive indicated in the past that I'm sympathetic to adding an *Icon* as a decoration to the existing Button class; but it's not a high priority for me personally.
2. OptionContainer on iOS/Android. "If I had an image button, I could implement my own Tab control, because Toga doesn't have one". iOS and Android both have native Tab navigation controls, and we're not going to add the ability to have clickable images so that people can work around a missing widget. 
3. An "image gallery" widget, where you need to be able to click on an image to activate some functionality. This is the *only* case where I agree that a "clickable image" might be a viable solution - but even then, I'd be more inclined to add an *actual* image gallery widget so as to avoid ideas 1 and 2 becoming viable.
4. An "image editing" widget. Again, this *might* be a use case for a clickable image, except that you need much, much more than just click handling.

Toga's design philosophy is *explicitly* trying to ensure that Toga apps are apps that are compliant with platform design guidelines. To that end, we're willing to *not* implement features if it encourages better GUI design. Clickable labels and images are the most common manifestation of this.

Adding right click and long press events are another matter entirely. We already *do* support right click (usually labeled "alternate press" in the API, because it's not necessarily the right button), and we use long press in some contexts to indicate "activation". If you've got a specific place where you think a right click or long press event handler would be appropriate, we'll consider them - but again, we're not going to add handlers to widgets that *do not need them*.

**Reply by wanglifeng95** [reply]
Clickable images are seen everywhere on desktop (Mac, Windows, Linux) and mobile(IOS, Android) APPS, as well as on the web. Widgets can be right-click and visible anywhere on the desktop, and long press events can also be seen anywhere on the mobile end. Developers know they need them for ease of development, but you don't think they need them. Ha, just like when developers want to eat apples, you tell them to eat oranges and pears instead. Your idea is correct, and the functionality that others want is also normal.. It's okay. We also can directly call the local API to implement the desired function. This is the strength of Toga, which is commendable for being able to accommodate so many system platforms. Of course, seeing the improvements made by GTK4.0 to GTK3.0, I also find that you are willing to upgrade the Linux API from GTK3 to GTK4.0 and I am looking forward to it.

**Reply by l-o-u-i-s-g** [reply]
freakboy3742   just look at the main apps:   Youtube, Instagram,  Facebook, Twitter  they all have images that can be clickable, buttons with icons and labels,  buttons with images and labels,  clickable listview items that may contain images, labels, icons  etc ..

**Reply by freakboy3742** [reply]
I don't know how many times I can repeat the same thing.

There is a **fundamental philosophical difference** between an **Image** (arbitrary size, arbitrary aspect ratio, full color spectrum) and an **Icon** (generally small, 1:1 aspect ratio, generally low color or single color, often provided in multiple resolutions). Toga has ``toga.Icon`` and ``toga.Image`` for this very specific reason.

I **AGREE** that a button with an **ICON** is a valid use case. I **SUPPORT** the idea of adding **ICONS** to Buttons. It's not a high priority for me personally, but if someone wants to add support for defining a button that uses an Icon, I'll review it.

I **DO NOT** support the idea of adding an ``on_click`` handle to **IMAGES**. This is a *fundamentally* different request, for the reasons that I've described, *at length*, *many* times, including *in this thread*.

You haven't provided specific examples to support your assertion, just nebulously waved at "the main apps". However, I'll warrant that if you examine those examples, every one of them where there's an "image that is clickable", the "image" is either (a) actually an *icon*, (b) [one of the examples I described above](https://github.com/beeware/toga/discussions/2141#discussioncomment-7206319) - most likely, an image gallery, or (c) an example of bad, undiscoverable UX.

**Reply by l-o-u-i-s-g** [reply]
For example on Facebook and twitter when you click on your picture  it takes you to your account profile. This is a picture  (of yourself)  so not an Icon. If you click on a picture in your wall then it takes you to a page where you see the picture bigger and the comments. On YouTube when you click on a picture  it starts a  video.  Note that there are other objects around, the picture. labels etc.  so a gallery widget would be too limiting.   All of them have a clickable picture of their respective brand that will navigate you to another view.  In twitter at the bottom you have your picture with text, ( yout name / id)  that you can click to then select to logout etc..  They all have  menu items  with a clickable  icon and text buttons. So I guess the best way to solve all cases would be  a box inside a button.  Then you can put whatever you want inside that box. Would solve all the cases. Don't know if that's feasible for all platforms thought.

**Reply by freakboy3742** [reply]
I'd argue the example of a profile picture on Facebook/Twitter *is* an icon - albeit a very specialised one. The critical detail: the image *must* be small and square. The only way that it varies here is that it's colourful, unlike most icons.

The YouTube example is *literally* a gallery. The existence of annotations and "extra text" is irrelevant; a photo gallery can also has text annotations - the name of the photo file, date, etc; a file manager is a "gallery of files", represented either by an icon, or a preview of the file, annotated with the file name. Annotations around the objects in a gallery doesn't change the fact that it's a gallery of items, where the image is a representation of the thing you're selecting.

The phrase "menu items with a clickable icon and text button" is baffling to me as a counterargument. Firstly, you explicitly describe it as using *an icon*, making my own point. Secondly, the broader use case is a navigation widget, not a button. Just because you can click on it doesn't mean it's a button.

And no - putting arbitrary items in a clickable box *doesn't* "solve all cases", because one of the cases is preventing developers from creating undiscoverable user interfaces with "buttons" that users can't work out are clickable because the UI doesn't provide any hints to that effect.

**Reply by l-o-u-i-s-g** [reply]
Ok so you would only differentiate and image from an  Icon just by restricting its shape and size?  I'm used to having icons as being objects that come from the framework itself.   So in flutter  it looks like this when you  want an icon:  Icon(Icons.arrow_forward ) and if you want an image then you have to specify a source, since its a file that the programmer adds to  its source so something like this: Image.asset('path/my_image.png').  

For the menu items I was simply making  the case for the combination of  an Icon and  Text.  

Let me show you a few code examples of how its done in Flutter: 

A button with an Icon: 
```
return MaterialButton(
  height: 50.0,
  child:Icon(Icons.add), 
  onPressed: () async {}
);
```
A button with an Image:
```
return MaterialButton(
  height: 50.0,
  child: Image.asset('images/snoopy.png'),
  onPressed: () async {}
);
```
A button with a combination of image or icon and text:
```
return MaterialButton(
  height: 50.0,
  child: Row(
    children: <Widget>[
      Icon(Icons.add), 
      Text("Add here"),
    ],
  ),
  onPressed: () async {}
);
```
In the last example  the child  can also be a column and you can put basically any other widgets in there. That's where I was  getting the idea of putting a box inside a button.
 
I guess you could be making specific widget for each user cases like a gallery  a menu item  etc .. but if you become too specific it will be limiting  and much more work to maintain, no?

**Reply by l-o-u-i-s-g** [reply]
Flutter also has  InkWell  to make anything respond  to touch and handle more complex examples.  I used a button in the examples for simplicity. https://api.flutter.dev/flutter/material/InkWell-class.html    
Perhaps you would rather go that route.  Instead of being a button,  simply an  "Event" entity  to which you can add any widgets.

**Reply by freakboy3742** [reply]
> Ok so you would only differentiate and image from an Icon just by restricting its shape and size? 

At least at present, there's no "built in" icons in Toga, but that's definitely something I would like to see change in time. 

However, even in that eventuality, there needs to be a way to define an icon that the system *doesn't* provide. It's impossible for a GUI framework to provide *every* icon for *every* purpose. What does an app developer do when a new social media platform emerges, and they want to support it, but their GUI framework doesn't provide an icon representing that platform?

> For the menu items I was simply making the case for the combination of an Icon and Text.

Sure - and I'm making the case that it's not "text and an icon" - *it's a menu*. That's a function that can be represented in an abstract way, and the way that abstract function is implemented need not be the same "text and an icon" on every platform. Compare the way that a phone renders Twitter and the way a desktop browser renders Twitter.

> Let me show you a few code examples of how its done in Flutter:

Allow me to assure you that my resistance here has nothing to do with feasibility, API design, or familiarity with GUI toolkits. I'm very much aware how it *can* be done in *many* other frameworks. Completely aside from the fact that Toga is wrapper around 6 different GUI (and 1 TUI) frameworks, I've been various using GUI APIs for over 30 years at this point.

The implementation, if it were to be added to Toga, would be *trivial*. However, it would also undermine the design goals of Toga as a framework.

> I guess you could be making specific widget for each user cases like a gallery a menu item etc .. but if you become too specific it will be limiting and much more work to maintain, no?

That's *exactly* the case I'm making.

Toga has [explicitly stated as a design goal](https://toga.readthedocs.io/en/stable/background/project/philosophy.html#abstract-the-broad-concepts) that we wrap abstract concepts, not specific widgets. Yes, this makes *some* use cases more difficult; but the benefit that comes from that is that apps are *always* well behaved and HIG-compliant examples of native apps. This is a design decision that has been made *actively*, with full awareness of the consequences.

(FWIW - this is also part of the reason [why the name of the framework is Toga](https://toga.readthedocs.io/en/stable/background/project/faq.html).

As for being "more work to maintain" - you're neglecting the role of third-party contributions to this space. In the long term, I'd consider it a failure if there isn't an ecosystem of useful, complex, third-party widgets encompassing interesting use cases. I'd also consider it a failure if the best of those don't find their way into Toga's core once they've proven robust and of general interest.

For that matter, if someone *really* wants a "clickable image", they *could* implement it *right now* as a third party widget. It wouldn't even be that much work. All I'm saying is that I'm not going to merge it into Toga as an official widget.

**Reply by l-o-u-i-s-g** [reply]
ok so is the confusion simply a poor choice of word "button" from my part?  as this implies a more specific  widget  with a single purpose? But do you agree  that Toga needs a way  for developers to create other clickable widgets without the need to create code for each platforms, IOS, osX, android, linux etc?   Correct me if I'm wrong  but I believe that right now if I wanted to make a gallery widget  or a  menu_item widget  I would need to write a version for each platforms?

**Reply by freakboy3742** [reply]
You're correct that to build a gallery widget at present you would currently need to write multiple platform-specific implementations.

I don't agree that Toga *necessarily* needs a way to create "clickable widgets". Toga's API intentionally doesn't expose a set of drawing primitives. It's a set of high level end-user use cases. That's deliberate, because it gives the backend implementations the flexibility to find the most appropriate representation of each high level use case. It also means we don't get into the weeds of trying to find cross-platform abstractions of low level concepts, which has plenty of potential to result in bad UX of apps when misused or abused, and results in slightly worse app performance (because every abstraction, by definition, introduces complexity).

Does this make it harder to implement a "gallery" widget? Yes. However - it also means we don't end up with Toga apps in the wild that have bad UX. From a broad project perspective - I'm willing to accept the compromise of having *less* features in the short term if it means the public interface for the features we *do* have are well designed, and yield a good cross-platform, HIG-compliant user experience.

Would it be nice to have primitives to make writing new widgets easier? Possibly. However, as a data point - almost without exception, every request for "clickable images" has been a request for one of 4 things:

1. "Pretty buttons", because "I don't like the macOS/GTK/Windows style of buttons".
2. "Buttons so I an implement a tab bar on mobile"
3. A image gallery
4. A high level app navigation control

(1) is *just plain wrong*, and I will die on this hill. Consistency of UI elements within the OS is a *feature*. It aids usability. Apps that decide they know better and restyle key UI elements annoy me beyond words, and I'm not going to support violating UI guidelines. If you don't agree with this design decision, you are looking for a different GUI toolkit. My immediate suggestion would be Kivy, because it's the diametric opposite of Toga's design - *no* widget adheres to the platform HIG, because they're all being drawn from primitives.

(2) is actually a request for an OptionContainer widget. If the amount of effort that had been spent furiously arguing for "clickable images" to support this use case, had instead been directed towards implementing OptionContainer on Android, we'd have it by now. The good news is that this widget exists on desktop platforms, and it's on our roadmap to implement towards the end of the year. 

(3) and (4) are reasonable use cases, and I would like to see them both resolved. A "clickable image" would be, at best, a bandaid solution.

I'd much rather spend the effort fixing the actual problems and functionality gaps, rather than compromise the design of the entire framework to support a couple of use cases.

**Reply by l-o-u-i-s-g** [reply]
After reflection I think you are going against your own philosophy.  You state that your goal is to stay as close to the feel of the native apps. Native apps do have  buttons with images/icons  and labels. By suggesting that instead  we create  other objects  like galleries  and menu items  etc.. you are moving away from  the native way of doing things.  For example  If one simply wanted a logout button with his picture and a label  he now has to create  a logout widget  where all he really needed was a button.  In the end  all these extra widgets that you suggest we create   will  end up being just a button  in the native codes.

**Reply by freakboy3742** [reply]
I assure you that I'm not being inconsistent. You're missing the point.

Providing wrappers around primitives is *not* "moving closer to the native way of doing things". It's *objectively* less native. Consider a widget like [DetailedList](https://toga.readthedocs.io/en/stable/reference/api/widgets/detailedlist.html). Using your approach, we would/could build a widget from images, labels, and scrollbars. That's not what Toga does. It wraps NSTableView on macOS, and UITableViewController on iOS, and GTK.ListBox on GTK - each of which provides a rich, fully native representation of the DetailedList concept. And on each platform, the implementation is *slightly different*, because the local platform expectations are different. On macOS, you use right click context menus to active functionality. On iOS, you use swipe left/right. Rinse and repeat for a dozen subtle differences between widgets on every platform. Toga's approach yields a user experience that is consistent with native expectations on every platform.

There isn't currently a Gallery widget. I am saying that I'd rather see a single, well-built gallery widget that uses native API features and GUI conventions than a dozen individual attempts at cobbling together a Gallery from a collection of "clickable image" widgets. I want to see a *native* representation. And, to that end, I'm *encouraging* that outcome by *not* providing primitives like "on_click handlers on image".

Are there examples where the desired functionality *might* ultimately be implemented by an on_click handler on an image? Possibly. Does that mean that there is a need for a clickable image widget? No. It means there are high level use cases that haven't been captured yet.
